module/vision/VisualMesh/src/VisualMesh.cpp

```diff
@@ -0,0 +1,174 @@
+
+                        for (const auto& v : l) {
+                            weight.push_back(v.as<float>());
+                        }
+                    }
+
+                    // Copy across our biases
+                    for (const auto& v : layer["biases"]) {
+                        net_layer.second.push_back(v.as<float>());
+                    }
+                }
+            }
+
+            // Make the classifier
+            classifier = mesh.make_classifier(network);
+
+            log("Finished loading visual mesh");
+        });
+
+        on<Trigger<Image>, With<CameraParameters>, Buffer<4>>().then([this](const Image& img,
+                                                                            const CameraParameters& cam) {
+            // Get our camera to world matrix
+            Eigen::Affine3f Hcw(img.Hcw.cast<float>());
+
+            // Transpose and store in our row major array
+            std::array<std::array<float, 4>, 4> Hoc;
+            Eigen::Map<Eigen::Matrix<float, 4, 4, Eigen::RowMajor>>(Hoc[0].data()) = Hcw.inverse().matrix();
+
+            // TODO: Un-hardcode
+            mesh::VisualMesh<float>::Lens lens;
+            switch (cam.lens.value) {
+                case CameraParameters::LensType::RADIAL:
+                    lens.projection   = mesh::VisualMesh<float>::Lens::EQUIDISTANT;
+                    lens.focal_length = 1.0 / cam.radial.radiansPerPixel;
+                    break;
+                case CameraParameters::LensType::PINHOLE:
+                    lens.projection   = mesh::VisualMesh<float>::Lens::RECTILINEAR;
+                    lens.focal_length = cam.pinhole.focalLengthPixels;
+                    break;
+            }
+            lens.dimensions = {{int(img.dimensions[0]), int(img.dimensions[1])}};
+            lens.fov        = cam.FOV.x();
+
+            auto results = classifier(img.data.data(), mesh::VisualMesh<float>::FOURCC(img.format), Hoc, lens);
+
+            // Get the mesh that was used so we can make our message
+            const auto& m = mesh.height(Hoc[2][3]);
+
+            auto msg = std::make_unique<VisualMeshMsg>();
+
+            // Pass through our camera id
+            msg->camera_id = img.camera_id;
+
+            // Add our description
+            for (const auto& r : m.rows) {
+                msg->mesh.emplace_back(r.phi, r.end - r.begin);
+            }
+
+            // Add our indices
+            msg->indices = std::move(results.global_indices);
+
+            // Add our neighbourhood
+            msg->neighbourhood.reserve(results.neighbourhood.size());
+            for (const auto& n : results.neighbourhood) {
+                msg->neighbourhood.emplace_back(Eigen::Map<const Eigen::Matrix<int, 6, 1, Eigen::DontAlign>>(n.data()));
+            }
+
+            // Add our coordinates
+            msg->coordinates = results.pixel_coordinates.as<Eigen::Matrix<int, 2, 1, Eigen::DontAlign>>();
```

Comment:
This is wrong here, you're basically doing a reinterpret cast on some floats into ints. That won't give your correct pixel coordinates.
it should be `results.pixel_coordinates.as<Eigen::Matrix<float, 2, 1, Eigen::DontAlign>>();`
but then you'll need to either make msg->coordinates floats or round the output